src/compare-tree/index.js

```diff
@@ -0,0 +1,197 @@
+   * @param {String} keyPaths A list of key paths to trigger the callback
+   * @param {Function} callback Function to execute when a key path changes
+   * @param {*} [scope] Optional scope to invoke the function with
+   */
+  on (keyPaths, callback, scope) {
+    let dependencies = getKeyPaths(keyPaths)
+    let id = Query.getId(keyPaths)
+
+    let query = this.addQuery(id, dependencies)
+
+    dependencies.map(this.track, this)
+                .forEach(node => node.connect(query))
+
+    query.on('change', callback, scope)
+
+    return query
+  },
+
+  /**
+   * Remove a subscription created by .on()
+   * @public
+   * @param {String} keyPaths A list of key paths for the subscription
+   * @param {Function} callback Associated callback function
+   * @param {*} [scope] Associated scope
+   */
+  off (keyPaths, callback, scope) {
+    let dependencies = getKeyPaths(keyPaths)
+    let id = Query.getId(keyPaths)
+
+    let query = this.nodes[id]
+
+    if (query) {
+      query.off('change', callback, scope)
+
+      if (query._events <= 0) {
+        this.prune(query, dependencies)
+      }
+    }
+  },
+
+  /**
+   * Compare a new snapshot to the last snapshot, triggering event
+   * subscriptions along the way.
+   * @public
+   * @param {*} state - New snapshot of state
+   */
+  update (snapshot) {
+    let last = this.snapshot
+    this.snapshot = snapshot
+
+    let root = this.nodes[ROOT_KEY]
+
+    if (root) {
+      this.scan(root, last, snapshot)
+    }
+  },
+
+  /**
+   * Add a node to the tree if it has not otherwise been added.
+   * @private
+   * @param {String} id Identifier for the node.
+   * @param {String} key Each node represents a key in a nested object.
+   */
+  addNode (id, key) {
+    if (!this.nodes[id]) {
+      this.nodes[id] = new Node(id, key)
+    }
+
+    return this.nodes[id]
+  },
+
+  /**
+   * Add a query to the tree if it has not otherwise been
+   * added. Queries are leaf nodes responsible for managing
+   * subscriptions.
+   * @private
+   * @param {String} id Identifier for the node.
+   * @param {String} keyPaths Each query tracks a list of key paths
+   */
+  addQuery (id, keyPaths) {
+    if (!this.nodes[id]) {
+      this.nodes[id] = new Query(id, keyPaths)
+    }
+
+    return this.nodes[id]
+  },
+
+  /**
+   * Remove a node, then remove parents if they have no more edges.
+   * @private
+   * @param {String} node Starting node
+   * @param {Array<String>} keyPaths determines the pathway to traverse
+   */
+  prune (node, keyPaths) {
+    for (var i = 0, len = keyPaths.length; i < len; i++) {
+      let key = getKeyString(keyPaths[i])
+      let parent = this.nodes[key]
+
+      if (parent) {
+        let index = parent.edges.indexOf(node)
+
+        if (~index) {
+          parent.edges.splice(index, 1)
+        }
+
+        if (parent.edges.length <= 0) {
+          this.prune(parent, [castPath(parent.id).slice(0, -1)])
+        }
+      }
+    }
+
+    delete this.nodes[node.id]
```

Comment:
This could be way shorter and easier to parse if we just sent in a `parent` property on Node and Query. Trouble is that Query can have multiple parents, but maybe it could be restructured.

Essentially I want:

```javascript
while (node) {
  if (node.parent) {
    node.parent.disconnect(node)
  }

  if (node.edges.length === 0) {
      this.nodes.delete[node.id]
  }

  node = node.parent
}
```

This function is responsible for cleaning up a tree so that it only contains the nodes for all current queries. So if we remove a query, and there are no longer any handlers, we want to "prune" the branch of key paths so that only the key paths we care about are maintained.

Thinking more on this, `Query` tracks all of its key paths, and we never remove a node outright, always as the product of removing a query. Maybe we could just traverse the key paths for the query.